drivers/crypto/crypto_mchp_sha_g1.c

```diff
@@ -0,0 +1,348 @@
+/*
+ * Copyright (C) 2025 Microchip Technology Inc. and its subsidiaries
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ */
+
+#define DT_DRV_COMPAT microchip_sha_g1_crypto
+
+#define LOG_LEVEL CONFIG_CRYPTO_LOG_LEVEL
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(cypto_mchp_sha_g1);
+
+#include <zephyr/crypto/crypto.h>
+#include <zephyr/device.h>
+#include <zephyr/drivers/clock_control/mchp_sam_pmc.h>
+#include <zephyr/init.h>
+#include <zephyr/kernel.h>
+#include <zephyr/sys/atomic.h>
+#include <zephyr/sys/util.h>
+
+#define MCHP_SHA_CAPS_SUPPORT (CAP_INPLACE_OPS | CAP_SEPARATE_IO_BUFS | CAP_SYNC_OPS)
+
+#define FIRST_WORD_4_PADDING_EMPTY_MSG 0x000000080
+
+struct crypto_mchp_sha_cfg {
+	sha_registers_t *const regs;
+	const struct sam_clk_cfg clock_cfg;
+};
+
+struct crypto_mchp_sha_data {
+	struct k_sem device_sem;
+};
+
+struct crypto_mchp_sha_algo_cfg {
+	enum hash_algo algo;
+	uint32_t sha_mr_algo;
+	uint8_t dgst_len;
+	uint8_t block_size;
+};
+
+struct crypto_mchp_sha_session {
+	struct crypto_mchp_sha_algo_cfg const *algo_cfg;
+	bool in_use;
+};
+
+static const struct crypto_mchp_sha_algo_cfg mchp_sha_algo_cfgs[4] = {
+	{CRYPTO_HASH_ALGO_SHA224, SHA_MR_ALGO_SHA224, 224 / BITS_PER_BYTE, 64},
+	{CRYPTO_HASH_ALGO_SHA256, SHA_MR_ALGO_SHA256, 256 / BITS_PER_BYTE, 64},
+	{CRYPTO_HASH_ALGO_SHA384, SHA_MR_ALGO_SHA384, 384 / BITS_PER_BYTE, 128},
+	{CRYPTO_HASH_ALGO_SHA512, SHA_MR_ALGO_SHA512, 512 / BITS_PER_BYTE, 128},
+};
+
+static struct crypto_mchp_sha_session mchp_sha_sessions[2];
+static struct k_sem mchp_sha_session_sem;
+
+static inline void mchp_sha_set_input(sha_registers_t *const regs, const unsigned char *data,
+				      int len)
+{
+	const uint32_t *data_u32 = (const uint32_t *)data;
+	uint32_t i = 0, v;
+
+	while (len >= sizeof(uint32_t)) {
+		v = data_u32[i];
+		if (i < ARRAY_SIZE(regs->SHA_IDATAR)) {
+			regs->SHA_IDATAR[i] = v;
+		} else {
+			regs->SHA_IODATAR[i - ARRAY_SIZE(regs->SHA_IDATAR)] = v;
+		}
+		i++;
+		len -= sizeof(uint32_t);
+	}
+
+	if (len > 0) {
+		v = data[i * sizeof(uint32_t) + 0];
+		if (len > 1) {
+			v |= (uint32_t)data[i * sizeof(uint32_t) + 1] << 8;
+		}
+		if (len > 2) {
+			v |= (uint32_t)data[i * sizeof(uint32_t) + 2] << 16;
+		}
+		if (i < ARRAY_SIZE(regs->SHA_IDATAR)) {
+			regs->SHA_IDATAR[i] = v;
+		} else {
+			regs->SHA_IODATAR[i - ARRAY_SIZE(regs->SHA_IDATAR)] = v;
+		}
+	}
+}
+
+static inline void mchp_sha_get_output(sha_registers_t *const regs, unsigned char *data, int len)
+{
+	uint32_t *data_u32 = (uint32_t *)data;
+	uint32_t i;
+
+	for (i = 0; i < (len / sizeof(uint32_t)) && i < ARRAY_SIZE(regs->SHA_IODATAR); i++) {
+		data_u32[i] = regs->SHA_IODATAR[i];
+	}
+}
+
+static int mchp_sha_wait_data_rdy(sha_registers_t *const regs)
+{
+	uint32_t timeout = 1;
+
+	while ((regs->SHA_ISR & SHA_ISR_DATRDY_Msk) == 0) {
+		if (timeout-- == 0) {
+			LOG_ERR("MCHP SHA wait data ready timeout");
+			return -ETIMEDOUT;
+		}
+		k_sleep(K_USEC(5));
```

Comment:
I may be missing something but this looks a completely arbitrary value (and too small). k_sleep is quantized to ticks and the value that you will sleep is much likely way bigger than that